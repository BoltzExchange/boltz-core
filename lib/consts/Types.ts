import { BIP32Interface } from 'bip32';
import { ECPairInterface } from 'ecpair';
import { TxOutput } from 'bitcoinjs-lib';
import { OutputType } from './Enums';

export type Error = {
  message: string;
  code: string;
};

export type ScriptElement = Buffer | number | string;

export type TransactionOutput = {
  txHash: Buffer;
  vout: number;
  type: OutputType;
} & TxOutput;

export type RefundDetails = TransactionOutput & {
  keys: ECPairInterface | BIP32Interface;
  redeemScript: Buffer;
};

export type ClaimDetails = RefundDetails & {
  preimage: Buffer;
};
