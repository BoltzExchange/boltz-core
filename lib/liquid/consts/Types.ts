import { Transaction, TxOutput } from 'liquidjs-lib';
import { RefundDetails } from '../../consts/Types';

export type LiquidRefundDetails = RefundDetails &
  TxOutput & {
    legacyTx?: Transaction;
    blindingPrivKey?: Buffer;
  };

export type LiquidClaimDetails = LiquidRefundDetails & {
  preimage: Buffer;
};
