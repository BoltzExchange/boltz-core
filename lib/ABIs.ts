import ERC20ABI from '../out/ERC20.sol/ERC20.json';
import ERC20SwapABI from '../out/ERC20Swap.sol/ERC20Swap.json';
import EtherSwapABI from '../out/EtherSwap.sol/EtherSwap.json';
import RouterABI from '../out/Router.sol/Router.json';

export const ContractABIs = {
  ERC20: ERC20ABI.abi,
  ERC20Swap: ERC20SwapABI.abi,
  EtherSwap: EtherSwapABI.abi,
  Router: RouterABI.abi,
} as const;
