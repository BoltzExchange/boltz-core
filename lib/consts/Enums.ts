export enum OutputType {
  Bech32 = 0,
  Compatibility = 1,
  Legacy = 2,
  Taproot = 3,
}
