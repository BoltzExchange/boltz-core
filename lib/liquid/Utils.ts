import { confidential, TxOutput } from 'liquidjs-lib';
import { confidentialLiquid } from './Init';

export const getOutputValue = (
  output: TxOutput & {
    blindingPrivKey?: Buffer;
  },
): number => {
  return output.blindingPrivKey
    ? Number(
        confidentialLiquid.unblindOutputWithKey(output, output.blindingPrivKey)
          .value,
      )
    : confidential.confidentialValueToSatoshi(output.value);
};
