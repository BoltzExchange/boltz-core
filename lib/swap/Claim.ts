/**
 * This file is based on the repository github.com/submarineswaps/swaps-service created by Alex Bosworth
 */

import * as bip65 from 'bip65';
import ops from '@boltz/bitcoin-ops';
import * as varuint from 'varuint-bitcoin';
import { crypto, script, Transaction } from 'bitcoinjs-lib';
import Errors from '../consts/Errors';
import { OutputType } from '../consts/Enums';
import { ClaimDetails } from '../consts/Types';
import { estimateFee, Input } from '../FeeCalculator';
import { encodeSignature, getOutputScriptType, scriptBuffersToScript } from './SwapUtils';

/**
 * Claim swaps
 *
 * @param utxos UTXOs that should be claimed or refunded
 * @param destinationScript the output script to which the funds should be sent
 * @param feePerByte how many satoshis per vbyte should be paid as fee
 * @param isRbf whether the transaction should signal full Replace-by-Fee
 * @param timeoutBlockHeight locktime of the transaction; only needed if the transaction is a refund
 */
export const constructClaimTransaction = (
  utxos: ClaimDetails[],
  destinationScript: Buffer,
  feePerByte: number,
  isRbf = true,
  timeoutBlockHeight?: number,
): Transaction => {
  for (const input of utxos) {
    if (input.type === OutputType.Taproot) {
      throw Errors.TAPROOT_NOT_SUPPORTED;
    }
  }

  const tx = new Transaction();

  // Refund transactions are just like claim ones and therefore this method
  // is also used for refunds. The locktime of refund transactions has to be
  // after the timelock of the UTXO is expired
  if (timeoutBlockHeight) {
    tx.locktime = bip65.encode({ blocks: timeoutBlockHeight });
  }

  // The sum of the values of all UTXOs that should be claimed or refunded
  let utxoValueSum = 0;
  const feeInputs: Input[] = [];

  utxos.forEach((utxo) => {
    utxoValueSum += utxo.value;
    feeInputs.push({ type: utxo.type, swapDetails: utxo });

    // Add the swap as input to the transaction
    //
    // RBF reference: https://github.com/bitcoin/bips/blob/master/bip-0125.mediawiki#summary
    tx.addInput(utxo.txHash, utxo.vout, isRbf ? 0xfffffffd : 0xffffffff);
  });

  // Estimate the fee for the transaction
  const fee = estimateFee(feePerByte, feeInputs, [getOutputScriptType(destinationScript)!]);

  // Send the sum of the UTXOs minus the estimated fee to the destination address
  tx.addOutput(destinationScript, utxoValueSum - fee);

  utxos.forEach((utxo, index) => {
    switch (utxo.type) {
      // Construct and sign the input scripts for P2SH inputs
      case OutputType.Legacy: {
        const sigHash = tx.hashForSignature(index, utxo.redeemScript, Transaction.SIGHASH_ALL);
        const signature = utxo.keys.sign(sigHash);

        const inputScript = [
          encodeSignature(Transaction.SIGHASH_ALL, signature),
          utxo.preimage,
          ops.OP_PUSHDATA1,
          utxo.redeemScript,
        ];

        tx.setInputScript(index, scriptBuffersToScript(inputScript));
        break;
      }

      // Construct the nested redeem script for nested SegWit inputs
      case OutputType.Compatibility: {
        const nestedScript = [
          varuint.encode(ops.OP_0).toString('hex'),
          crypto.sha256(utxo.redeemScript),
        ];

        const nested = scriptBuffersToScript(nestedScript);

        tx.setInputScript(index, scriptBuffersToScript([ nested ]));
        break;
      }
    }

    // Construct and sign the witness for (nested) SegWit inputs
    if (utxo.type !== OutputType.Legacy) {
      const sigHash = tx.hashForWitnessV0(index, utxo.redeemScript, utxo.value, Transaction.SIGHASH_ALL);
      const signature = script.signature.encode(utxo.keys.sign(sigHash), Transaction.SIGHASH_ALL);

      tx.setWitness(index, [
        signature,
        utxo.preimage,
        utxo.redeemScript,
      ]);
    }
  });

  return tx;
};
