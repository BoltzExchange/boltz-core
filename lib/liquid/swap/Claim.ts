import ops from '@boltz/bitcoin-ops';
import {
  Blinder,
  Creator,
  CreatorInput,
  CreatorOutput,
  crypto,
  Extractor,
  Finalizer,
  Pset,
  script,
  Signer,
  Transaction,
  Updater,
  witnessStackToScriptWitness,
  ZKPGenerator,
  ZKPValidator,
} from 'liquidjs-lib';
import { Network } from 'liquidjs-lib/src/networks';
import { toXOnly } from 'bitcoinjs-lib/src/psbt/bip371';
import { reverseBuffer, varuint } from 'liquidjs-lib/src/bufferutils';
import { ecpair, secp } from '../init';
import Networks from '../consts/Networks';
import { getOutputValue } from '../Utils';
import { getHexString } from '../../Utils';
import { OutputType } from '../../consts/Enums';
import { validateInputs } from '../../swap/Claim';
import { LiquidClaimDetails } from '../consts/Types';
import { scriptBuffersToScript } from '../../swap/SwapUtils';
import { createControlBlock, tapLeafHash, toHashTree } from './TaprooUtils';

const dummyWitness = Buffer.from('0x21', 'hex');

const getSighashType = (type: OutputType) =>
  type === OutputType.Taproot
    ? Transaction.SIGHASH_DEFAULT
    : Transaction.SIGHASH_ALL;

/**
 * Claim swaps
 *
 * @param utxos UTXOs that should be claimed or refunded
 * @param destinationScript the output script to which the funds should be sent
 * @param fee how many satoshis should be paid as fee
 * @param isRbf whether the transaction should signal full Replace-by-Fee
 * @param network network of the transaction
 * @param blindingKey blinding public key for the output; undefined if the output should not be blinded
 * @param timeoutBlockHeight locktime of the transaction; only needed if the transaction is a refund
 * @param isRefund whether the transaction is a refund or claim
 */
export const constructClaimTransaction = (
  utxos: LiquidClaimDetails[],
  destinationScript: Buffer,
  fee: number,
  isRbf = true,
  network: Network = Networks.liquidMainnet,
  blindingKey?: Buffer,
  timeoutBlockHeight?: number,
  isRefund = false,
): Transaction => {
  validateInputs(utxos);

  if (
    utxos.some(
      (utxo) =>
        (utxo.blindingPrivateKey === undefined) !==
        (utxos[0].blindingPrivateKey === undefined),
    )
  ) {
    throw 'all or none inputs have to be blinded';
  }

  if (
    utxos.some(
      (utxo) =>
        utxo.type !== OutputType.Taproot && utxo.type !== OutputType.Bech32,
    )
  ) {
    throw 'only Taproot or native SegWit inputs supported';
  }

  const pset = Creator.newPset();
  const updater = new Updater(pset);

  let utxoValueSum = BigInt(0);

  for (const [i, utxo] of utxos.entries()) {
    utxoValueSum += BigInt(getOutputValue(utxo));

    const txHash = Buffer.alloc(utxo.txHash.length);
    utxo.txHash.copy(txHash);
    const txid = getHexString(reverseBuffer(txHash));

    pset.addInput(
      new CreatorInput(
        txid,
        utxo.vout,
        isRbf ? 0xfffffffd : 0xffffffff,
        i == 0 ? timeoutBlockHeight : undefined,
      ).toPartialInput(),
    );
    updater.addInSighashType(i, getSighashType(utxo.type));

    if (utxo.type === OutputType.Legacy) {
      updater.addInNonWitnessUtxo(i, utxo.legacyTx!);
      updater.addInRedeemScript(i, utxo.redeemScript!);
    } else if (utxo.type === OutputType.Compatibility) {
      updater.addInNonWitnessUtxo(i, utxo.legacyTx!);
      updater.addInRedeemScript(
        i,
        scriptBuffersToScript([
          scriptBuffersToScript([
            varuint.encode(ops.OP_0).toString('hex'),
            crypto.sha256(utxo.redeemScript!),
          ]),
        ]),
      );
    }

    if (utxo.type !== OutputType.Legacy) {
      updater.addInWitnessUtxo(i, utxo);
      updater.addInWitnessScript(i, utxo.redeemScript!);
    }
  }

  updater.addOutputs([
    {
      script: destinationScript,
      blindingPublicKey: blindingKey,
      asset: network.assetHash,
      amount: Number(utxoValueSum - BigInt(fee)),
      blinderIndex: blindingKey !== undefined ? 0 : undefined,
    },
  ]);

  const addFeeOutput = () => {
    updater.addOutputs([
      {
        amount: fee,
        asset: network.assetHash,
      },
    ]);
  };

  if (utxos[0].blindingPrivateKey !== undefined) {
    // We have to have at least one blinded output if we are spending blinded coins,
    // so we add a small OP_RETURN
    if (blindingKey === undefined) {
      pset.addOutput(
        new CreatorOutput(
          network.assetHash,
          0,
          Buffer.of(ops.OP_RETURN),
          ecpair.makeRandom().publicKey,
          0,
        ).toPartialOutput(),
      );
    }

    addFeeOutput();

    blindPset(pset, utxos);
  } else {
    addFeeOutput();
  }

  const signer = new Signer(pset);

  const signatures: Buffer[] = [];

  for (const [i, utxo] of utxos.entries()) {
    if (utxo.type === OutputType.Taproot) {
      if (utxo.cooperative) {
        signatures.push(dummyWitness);
        continue;
      }

      const leafHash = tapLeafHash(
        isRefund ? utxo.swapTree!.refundLeaf : utxo.swapTree!.claimLeaf,
      );
      const signature = utxo.keys.signSchnorr(
        pset.getInputPreimage(
          i,
          getSighashType(utxo.type),
          network.genesisBlockHash,
          leafHash,
        ),
      );
      signatures.push(signature);
      signer.addSignature(
        i,
        {
          genesisBlockHash: network.genesisBlockHash,
          tapScriptSigs: [
            {
              signature: signature,
              pubkey: toXOnly(utxo.keys.publicKey),
              leafHash,
            },
          ],
        },
        Pset.SchnorrSigValidator(secp.ecc),
      );
    } else {
      const signature = script.signature.encode(
        utxo.keys.sign(pset.getInputPreimage(i, getSighashType(utxo.type))),
        getSighashType(utxo.type),
      );
      signatures.push(signature);

      signer.addSignature(
        i,
        {
          partialSig: {
            pubkey: utxo.keys.publicKey,
            signature,
          },
        },
        Pset.ECDSASigValidator(secp.ecc),
      );
    }
  }

  const finalizer = new Finalizer(pset);

  for (const [i, utxo] of utxos.entries()) {
    finalizer.finalizeInput(i, () => {
      const finals: {
        finalScriptSig?: Buffer;
        finalScriptWitness?: Buffer;
      } = {};

      if (utxo.type === OutputType.Legacy) {
        finals.finalScriptSig = scriptBuffersToScript([
          signatures[i],
          utxo.preimage,
          ops.OP_PUSHDATA1,
          utxo.redeemScript!,
        ]);
      } else if (utxo.type === OutputType.Compatibility) {
        finals.finalScriptSig = scriptBuffersToScript([
          scriptBuffersToScript([
            varuint.encode(ops.OP_0).toString('hex'),
            crypto.sha256(utxo.redeemScript!),
          ]),
        ]);
      }

      if (utxo.type === OutputType.Taproot) {
        if (utxo.cooperative) {
          // Add a dummy to allow for extraction
          finals.finalScriptWitness = witnessStackToScriptWitness([
            dummyWitness,
          ]);
        } else {
          const tapleaf = isRefund
            ? utxo.swapTree!.refundLeaf
            : utxo.swapTree!.claimLeaf;

          const witness = isRefund
            ? [signatures[i]]
            : [signatures[i], utxo.preimage];

          finals.finalScriptWitness = witnessStackToScriptWitness(
            witness.concat([
              tapleaf.output,
              createControlBlock(
                toHashTree(utxo.swapTree!.tree),
                tapleaf,
                utxo.internalKey!,
              ),
            ]),
          );
        }
      } else if (utxo.type !== OutputType.Legacy) {
        finals.finalScriptWitness = witnessStackToScriptWitness([
          signatures[i],
          utxo.preimage,
          utxo.redeemScript!,
        ]);
      }

      return finals;
    });
  }

  return Extractor.extract(pset);
};

const blindPset = (pset: Pset, utxos: LiquidClaimDetails[]) => {
  const zkpGenerator = new ZKPGenerator(
    secp as any,
    ZKPGenerator.WithBlindingKeysOfInputs(
      utxos.map((utxo) => utxo.blindingPrivateKey!),
    ),
  );
  const zkpValidator = new ZKPValidator(secp as any);
  const outputBlindingArgs = zkpGenerator.blindOutputs(
    pset,
    Pset.ECCKeysGenerator(secp.ecc),
  );

  const blinder = new Blinder(
    pset,
    zkpGenerator.unblindInputs(pset),
    zkpValidator,
    zkpGenerator,
  );

  blinder.blindLast({ outputBlindingArgs });
};
