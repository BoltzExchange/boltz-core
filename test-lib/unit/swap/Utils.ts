import Errors from '../../../lib/consts/Errors';
import { OutputType } from '../../../lib/consts/Enums';
import { p2wshOutput, p2shOutput, p2shP2wshOutput } from '../../../lib/swap/Scripts';

export const getScriptHashFunction = (type: OutputType): (scriptHex: Buffer) => Buffer => {
  switch (type) {
    case OutputType.Bech32: return p2wshOutput;
    case OutputType.Legacy: return p2shOutput;
    case OutputType.Compatibility: return p2shP2wshOutput;
    case OutputType.Taproot: throw Errors.TAPROOT_NOT_SUPPORTED;
  }
};
