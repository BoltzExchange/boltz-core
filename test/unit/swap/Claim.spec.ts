import { getHexBuffer } from '../../../lib/Utils';
import { OutputType } from '../../../lib/consts/Enums';
import { ClaimDetails } from '../../../lib/consts/Types';
import { claimDetails, claimDetailsMap } from './ClaimDetails';
import { constructClaimTransaction } from '../../../lib/swap/Claim';

describe('Claim', () => {
  const testClaim = (utxos: ClaimDetails[], fee: number) => {
    return constructClaimTransaction(
      utxos,
      getHexBuffer('00140000000000000000000000000000000000000000'),
      fee,
      false,
    );
  };

  test.each`
    type                        | fee    | name
    ${OutputType.Bech32}        | ${131} | ${'P2WSH'}
    ${OutputType.Compatibility} | ${166} | ${'P2SH nested P2WSH'}
    ${OutputType.Legacy}        | ${274} | ${'P2SH'}
  `(`should claim a $name swap`, ({ type, fee }) => {
    expect(
      testClaim([claimDetailsMap.get(type)!], fee).toHex(),
    ).toMatchSnapshot();
  });

  test('should claim multiple swaps in one transaction', () => {
    expect(testClaim(claimDetails, 490).toHex()).toMatchSnapshot();
  });
});
