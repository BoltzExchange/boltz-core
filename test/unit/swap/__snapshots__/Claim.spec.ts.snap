// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Claim should claim a P2SH nested P2WSH swap 1`] = `"01000000000101285d227e2823c679c224b4d562a9b5b5b7b927badd483df9f4225c6fc761d75400000000232200206f38b6ce82427d4df080a9833d06cc6c66ab816545c9fd4df50f9d1ca8430b9effffffff012a07000000000000160014000000000000000000000000000000000000000003483045022100ac9251c4d2e67f416d5ecb5e4033996fc7fd07c340349b634a0eb6013596dcdb02205ec8bbea731eec9977ab52e80fc1f90248960298db02b07d64b93e1378ef97ed0110b5b2dbb1f0663878ecbc20323b58b92c64a914a0738c92fde6361f09d28950c7bd0d2bf32b34be87632103be4a251dae719d565ce1d6a7a5787df99fc1ecc1f6e847567981a686f32abce167027802b1752103f7877d4ae985bb30b6f150ad6b6b9935c342432beed1a4781347b169c1e2417368ac00000000"`;

exports[`Claim should claim a P2SH swap 1`] = `"0100000001285d227e2823c679c224b4d562a9b5b5b7b927badd483df9f4225c6fc761d75400000000bf473044022026f3dae5c1c82ec82976acb8ff59b48c69940ae7c400c8faf8f583ea6d0f82f3022033cf14a6e7ccadc17df394ffa37fbc93d9218f409b038e2f6251b6e1dec655b30110b5b2dbb1f0663878ecbc20323b58b92c4c64a914a0738c92fde6361f09d28950c7bd0d2bf32b34be87632103be4a251dae719d565ce1d6a7a5787df99fc1ecc1f6e847567981a686f32abce167027802b1752103f7877d4ae985bb30b6f150ad6b6b9935c342432beed1a4781347b169c1e2417368acffffffff01be06000000000000160014000000000000000000000000000000000000000000000000"`;

exports[`Claim should claim a P2WSH swap 1`] = `"01000000000101285d227e2823c679c224b4d562a9b5b5b7b927badd483df9f4225c6fc761d7540000000000ffffffff014d070000000000001600140000000000000000000000000000000000000000034730440220656eb1b9e3d3a0af0af0c018f5d9038b64dd86041209102a0d7b4b84f845b75302203bf6c2057cc68abf6a84da806fec7984d773557de322eef1d1b82e4bb34792a80110b5b2dbb1f0663878ecbc20323b58b92c64a914a0738c92fde6361f09d28950c7bd0d2bf32b34be87632103be4a251dae719d565ce1d6a7a5787df99fc1ecc1f6e847567981a686f32abce167027802b1752103f7877d4ae985bb30b6f150ad6b6b9935c342432beed1a4781347b169c1e2417368ac00000000"`;

exports[`Claim should claim multiple swaps in one transaction 1`] = `"01000000000103285d227e2823c679c224b4d562a9b5b5b7b927badd483df9f4225c6fc761d7540000000000ffffffff285d227e2823c679c224b4d562a9b5b5b7b927badd483df9f4225c6fc761d75400000000bf473044022004bee4cf117e0d827224c1067c990a9d1d44597c2f71448a30b999c20c7caddd02206d9408eb4b50097538be1dbf0d835c72ca0f2bbfee85978f268d913f099ed38e0110b5b2dbb1f0663878ecbc20323b58b92c4c64a914a0738c92fde6361f09d28950c7bd0d2bf32b34be87632103be4a251dae719d565ce1d6a7a5787df99fc1ecc1f6e847567981a686f32abce167027802b1752103f7877d4ae985bb30b6f150ad6b6b9935c342432beed1a4781347b169c1e2417368acffffffff285d227e2823c679c224b4d562a9b5b5b7b927badd483df9f4225c6fc761d75400000000232200206f38b6ce82427d4df080a9833d06cc6c66ab816545c9fd4df50f9d1ca8430b9effffffff018615000000000000160014000000000000000000000000000000000000000003483045022100df1acb7342688501897dec3a9f0a484c431289e3fc43efbb9b8b310e9b50a26a02205751528b637811c9c5eb4cde667c889fb4d328db83ded869a419283212f7cac70110b5b2dbb1f0663878ecbc20323b58b92c64a914a0738c92fde6361f09d28950c7bd0d2bf32b34be87632103be4a251dae719d565ce1d6a7a5787df99fc1ecc1f6e847567981a686f32abce167027802b1752103f7877d4ae985bb30b6f150ad6b6b9935c342432beed1a4781347b169c1e2417368ac0003483045022100df1acb7342688501897dec3a9f0a484c431289e3fc43efbb9b8b310e9b50a26a02205751528b637811c9c5eb4cde667c889fb4d328db83ded869a419283212f7cac70110b5b2dbb1f0663878ecbc20323b58b92c64a914a0738c92fde6361f09d28950c7bd0d2bf32b34be87632103be4a251dae719d565ce1d6a7a5787df99fc1ecc1f6e847567981a686f32abce167027802b1752103f7877d4ae985bb30b6f150ad6b6b9935c342432beed1a4781347b169c1e2417368ac00000000"`;
